packages/expo/src/useOAuth.ts

```diff
@@ -0,0 +1,84 @@
+import { useSignIn, useSignUp } from '@clerk/clerk-react';
+import type { OAuthStrategy } from '@clerk/types';
+import * as AuthSession from 'expo-auth-session';
+import * as WebBrowser from 'expo-web-browser';
+
+export type UseOAuthFlowParams =
+  | {
+      strategy: OAuthStrategy;
+      redirectUrl?: string;
+    }
+  | OAuthStrategy;
+
+export function useOAuth(params: UseOAuthFlowParams) {
+  let strategy: OAuthStrategy;
+  let redirectUrl: string;
+
+  if (typeof params === 'string') {
+    strategy = params;
+  } else if (typeof params === 'object') {
+    strategy = params.strategy;
+    redirectUrl = params.redirectUrl || '';
+  }
+
+  const { signIn, setSession, isLoaded: isSignInLoaded } = useSignIn();
+  const { signUp, isLoaded: isSignUpLoaded } = useSignUp();
+
+  async function startOAuthFlow() {
+    if (!isSignInLoaded || !isSignUpLoaded) {
+      return {};
+    }
+
+    // Create a redirect url for the current platform and environment.
+    //
+    // This redirect URL needs to be whitelisted for your Clerk production instance via
+    // https://clerk.dev/docs/reference/backend-api/tag/Redirect-URLs#operation/CreateRedirectURL
+    //
+    // For more information go to:
+    // https://docs.expo.dev/versions/latest/sdk/auth-session/#authsessionmakeredirecturi
+    const oauthRedirectUrl =
+      redirectUrl ||
+      AuthSession.makeRedirectUri({
+        path: '/oauth-native-callback',
+      });
+
+    await signIn.create({ strategy, redirectUrl: oauthRedirectUrl });
+
+    const { externalVerificationRedirectURL } = signIn.firstFactorVerification;
+
+    const result = await WebBrowser.openAuthSessionAsync(externalVerificationRedirectURL!.toString(), oauthRedirectUrl);
+
+    // @ts-expect-error
+    const { type, url } = result || {};
+
+    // TODO: Check all the possible AuthSession results
+    // https://docs.expo.dev/versions/latest/sdk/auth-session/#returns-7
+    if (type !== 'success') {
+      throw 'Something went wrong during the OAuth flow. Try again.';
+    }
+
+    const params = new URL(url).searchParams;
```

Comment:
I'm asking because I'm not sure so please ignore if I'm completely off :) Is `URL` supported by all expo versions we support?

Reply:
Yes